Stubbed out Logging

RageCage64 opened the pull request:
Added the `zerolog` package to allow for logging throughout the application. Added support to configure a logger that can be injected, and injected the logger into the graphql server and a new request logging middleware.

`make test` Result:
![image](https://user-images.githubusercontent.com/6193285/126021975-6e516b3c-012f-4678-9ab1-4c3ca99bdc7f.png)

FrozenSake commented on middleware/logger.go:
```diff
@@ -0,0 +1,21 @@
+func MakeLoggingMiddlewareHandler(logger zerolog.Logger) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		handler := func(w http.ResponseWriter, r *http.Request) {
+			requestLogMessage := fmt.Sprintf("%s %s %s", r.RemoteAddr, r.URL, r.Method)
```
Should this have a lightweight message explaining where it is. Something like `Handling Request: %s %s %s`. My thought here is that once there are more logs, it will be unclear what this log is.

FrozenSake commented on middleware/logger.go:
```diff
@@ -0,0 +1,21 @@
+		handler := func(w http.ResponseWriter, r *http.Request) {
+			requestLogMessage := fmt.Sprintf("%s %s %s", r.RemoteAddr, r.URL, r.Method)
+			logger.Info().Msg(requestLogMessage)
+			next.ServeHTTP(w, r)
```
Would we want an after of `Request Complete` - I'm on the fence. It's not super necessary but could make a request chain easier to follow to termination.

Also, should the logger be passed down in a context?

BraydonKains replied:
The solution to have logs after the request were pretty involved; a lot of the solutions I saw involved wrapping the `http.ResponseWriter` in another struct, and that seemed pretty intense since I'm not sure yet what we'd even want to log there. 

I inject the logger into the server instead since I thought that made more sense, but maybe the context would be more useful.

FrozenSake replied:
Can't you just put a logger.Info() after the next() for an after-log? It may not have any value, and not worth the effort now if we don't get value-add, since  if it ever becomes important it's easy for the person who sees value.

Thoughts on that - if we inject it into the server, we can't pass it out if we call something else that's not this server. That's not necessarily a problem, especially since our expected footprint is small, and moving it to context later is not a huge lift because it's high on the chain.

FrozenSake commented on middleware/logger_test.go:
```diff
@@ -0,0 +1,38 @@
+
+	middlewareChain.ServeHTTP(w, req)
+
+	var log struct {
+		Level   string `json:"level"`
+		Message string `json:"message"`
```
zerologger attaches a time field, and I think it would be of value to verify it's coming back as not-nil. Something like

`assert.True(t, strings.Contains(log.Message, "time")`
`assert.Greater(t, 0, log.Time)`

BraydonKains replied:
zerolog actually doesn't pass a time field with the default `zerologger.New`, but I can make it do that and test it.

FrozenSake replied:
Oh, my bad. I saw 
```
//     log.Info().Msg("hello world")
//     // Output: {"time":1494567715,"level":"info","message":"hello world"}
```
in the docs and assumed it would do that. I would assume we want timestamps on our logs, personally, but that's not necessarily true?

zdylag replied:
I would basically always like a timestamp, yes. 👍

FrozenSake commented on app/graphql_server/main.go:
```diff
@@ -19,7 +23,81 @@ import (
 )
 
 func main() {
```
Is there a convention on where the `main()` func goes? I'm used to seeing it at the bottom of `main.go`, but I don't know if that's an actual convention, or just my brain drawing patterns.

BraydonKains replied:
The way I've always done it is `main()` at the top, and all other functions below it in the order they are called. Fine to change it if we want something different.

FrozenSake replied:
Eh, I'm indifferent then. Whatever we settle on we should probably put into Contributing.md though. Maybe get opinions from others on the backend?

zysim replied:
Couldn't care any less even if I tried.

FrozenSake reviewed (commented):
All go tests pass.
Go fmt gave me nothing important (I ran it in powershell so it wanted to do `LF->CRLF`)

Some questions, otherwise.

zdylag commented on middleware/logger.go:
```diff
@@ -0,0 +1,21 @@
+func MakeLoggingMiddlewareHandler(logger zerolog.Logger) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		handler := func(w http.ResponseWriter, r *http.Request) {
+			requestLogMessage := fmt.Sprintf("Handling Request: %s %s %s", r.RemoteAddr, r.URL, r.Method)
+			logger.Info().Msg(requestLogMessage)
```
Why not use `.Msgf(...)`?

zdylag reviewed (changes requested):
Why does this try to merge a 15MB binary file?

BraydonKains commented on app/graphql_server/main.go:
```diff
@@ -19,7 +23,81 @@ import (
+
+// Use the CLI flags passed in to construct a log configuration.
+func getLogConfigFromFlags() logConfig {
+	config := logConfig{
```
Synced offline and came up with a solution that will use environment variables instead.

BraydonKains commented on graphql_server/server.go:
```diff
@@ -13,9 +14,10 @@ import (
+	Users  []*data.User
+	Games  []*data.Game
+	Runs   []*data.Run
+	Logger zerolog.Logger
```
Synced offline and came up with a solution that will use a shared logging package with a globally exported member to avoid the future plumbing problem with this approach.

RageCage64 committed:
logger, middleware, app/graphql_server: see body

Added globally exported Logger in logger package, use it in middleware
and app/graphql_server

switching branches this commit will be amended

Makefile: Added run configurations with logging flags

middleware, logger: Added info to request log message

logger, middleware: added logger package and middleware

RageCage64 committed:
middleware: Check error value of json unmarshal

RageCage64 committed:
middleware: Use zerolog Msgf instead of a variable

FrozenSake reviewed (commented):
I'm not sure about the middleware application. It seems like we have to similarly call the logger anywhere downstream where we want it to log something else in this scenario? Basically this middleware doesn't inject a logger, it just logs the specific request's details.

I don't think that's necessarily a problem, but it's a choice we'd have to be clear on - if we have the chain of:
Metrics > Logger > Auth, we'd have to add a logging call directly to the global logger in Auth to log anything, right?

zysim commented on middleware/logger.go:
```diff
@@ -0,0 +1,13 @@
+
+// Make a logging middleware handler with the configured logger injected.
+func LoggingMiddleware(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	logger.Logger.Info().Msgf("Handling Request: %s %s %s", r.RemoteAddr, r.URL, r.Method)
```
How do we make sure that `logger.Logger` is defined at the point of calling this?

BraydonKains replied:
We don't necessarily make sure that `InitLogger` has been run. If I remember correctly though (it's been a while since I looked at this) if InitLogger hasn't been run it doesn't break or anything. It just won't have the settings we want; it will have the settings of the 0-value `zerolog.Logger`.
